src/pybind/mgr/orchestrator.py

```diff
@@ -0,0 +1,288 @@
+        if they proceed.
+        """
+        raise NotImplementedError()
+
+    def test_stateful_service_rule(self, service_type, stateful_service_spec,
+                                   placement_spec):
+        """
+        See add_stateful_service_rule.
+        """
+        raise NotImplementedError()
+
+    def remove_stateful_service_rule(self, service_type, id_):
+        """
+        This will remove the *rule* but not the services that were
+        created as a result.  Those should be converted into statically
+        placed services as if they had been created with add_stateful_service,
+        so that they can be removed with remove_stateless_service
+        if desired.
+        """
+        raise NotImplementedError()
+
+
+class UpgradeSpec(object):
+    # Request to orchestrator to initiate an upgrade to a particular
+    # version of Ceph
+    def __init__(self):
+        self.target_version = None
+
+
+class UpgradeStatusSpec(object):
+    # Orchestrator's report on what's going on with any ongoing upgrade
+    def __init__(self):
+        self.in_progress = False  # Is an upgrade underway?
+        self.services_complete = []  # Which daemon types are fully updated?
+        self.message = ""  # Freeform description
+
+
+class PlacementSpec(object):
+    """
+    For APIs that need to specify a node subset
+    """
+    def __init__(self):
+        self.label = None
+
+
+class StatefulServiceSpec(object):
+    """
+    Details of stateful service creation.
+
+    For OSDs this should map to what ceph-volume can do.
+    """
+
+    def __init__(self):
+        self.service_type = None  # osd, mon
+
+        self.format = None  # filestore, bluestore (meaningful for OSD only)
+
+        self.node = None  # name of a node
+
+        # Map role to device ID:
+        #   Valid keys: data (mon)
+        #   Valid keys: data, journal (filestore)
+        #   Valid keys: data, wal, db (bluestore)
+        self.block_devices = {}
+
+        # Arbitrary JSON-serializable object.
+        # Maybe your orchestrator knows how to do something
+        # special like encrypting drives
+        self.extended = {}
+
+
+class StatelessServiceSpec(object):
+    # Request to orchestrator for a group of stateless services
+    # such as MDS, RGW, nfs gateway, iscsi gateway
+    """
+    Details of stateless service creation.
+
+    This is *not* supposed to contain all the configuration
+    of the services: it's just supposed to be enough information to
+    execute the binaries.
+    """
+
+    def __init__(self):
+        self.placement = PlacementSpec()
+
+        # Give this set of statelss services a name: typically it would
+        # be the name of a CephFS pool, RGW zone, etc.  Must be unique
+        # within one ceph cluster.
+        self.name = ""
+
+        # Minimum and maximum number of service instances
+        self.min_size = 1
+        self.max_size = 1
```

Comment:
For salt and ansible, min_ and max_ sizes are probably irrelevant. I had an idea that it could be useful (I'm thinking of the UI) to know what options are/aren't available to the underlying system so options that aren't available are hidden from the user. In this case, I envision that the Orchestrator may have a Capabilities class like:

```py
def Capabilities(object):
    def __init__(self):
        self.supports_autoscaling = False
        # Below could be eliminated if the UI instead goes through a kind of self-test to
        # see which operations return NotImplemented, and then can gray-out unsupported
        # elements
        self.supports_stateless_services = False
```

If the UI sees that the Orchestrator doesn't support autoscaling, it may display only a single "size" field which will populate min_ and max_ with the same number.

Reply:
If an orchestrator doesn't natively understand how to schedule a number of services, I'd expect the ceph-mgr python code for that orchestrator to take care of it.  I'd imagine everyone is going to want the ability to have e.g. more than one RGW/MDS